<?xml version="1.0" encoding="UTF-8"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:property name="M_PI" value="3.1415926"/> 
  <xacro:property name="laser_min_range" value="0.1"/>
  <xacro:property name="laser_max_range" value="200.0"/>
  <xacro:property name="samples" value="40000"/>
  <xacro:property name="resolution" value="0.002"/>
  <xacro:property name="noise_mean" value="0.0"/>
  <xacro:property name="noise_stddev" value="0.01"/>

  <xacro:property name="mass" value="0.4" />
  <xacro:property name="length" value="0.1" />
  <xacro:property name="width" value="0.06" />
  <xacro:property name="height" value="0.06" />

  <xacro:macro name="box_inertia" params="m w h d">
    <inertial>
      <origin xyz="0 0 0" rpy="${pi/2} 0 ${pi/2}"/>
      <mass value="${m}"/>
      <inertia ixx="${(m/12) * (h*h + d*d)}" ixy="0.0" ixz="0.0" iyy="${(m/12) * (w*w + d*d)}" iyz="0.0" izz="${(m/12) * (w*w + h*h)}"/>
    </inertial>
  </xacro:macro>

  <xacro:macro name="mid360_harmonic" params="name:='' parent:= 'base_link' topic:='mid360' *origin ">

<link name="${name}">
<xacro:box_inertia m="${length}" w="${width}" d="${length}" h="${height}"/>
  <visual>
        <geometry>
            <box size="${length} ${width} ${height}"/>
        </geometry>
  </visual>
</link>

    <joint name="${name}_joint" type="fixed">
      <parent link="${parent}"/>
      <child link="${name}"/>
      <xacro:insert_block name="origin"/>
    </joint>

    <gazebo reference="${name}">
      <!-- Gazebo Harmonic Lidar Sensor -->
      <sensor name="${name}" type="gpu_lidar">
        <always_on>true</always_on>
        <update_rate>10</update_rate>
        <visualize>true</visualize>
        <lidar>
          <scan>
            <horizontal>
              <samples>360</samples>
              <resolution>1.0</resolution>
              <min_angle>0</min_angle>
              <max_angle>6.28318</max_angle>
            </horizontal>
            <vertical>
              <samples>32</samples>
              <resolution>1.0</resolution>
              <min_angle>-0.126</min_angle>
              <max_angle>0.964</max_angle>
            </vertical>
          </scan>
          <range>
            <min>${laser_min_range}</min>
            <max>${laser_max_range}</max>
            <resolution>0.001</resolution>
          </range>
          <noise>
            <type>gaussian</type>
            <mean>${noise_mean}</mean>
            <stddev>${noise_stddev}</stddev>
          </noise>
        </lidar>
        
        <!-- Gazebo Harmonic Plugin Configuration -->
        <plugin filename="libros2_livox.so" name="ros2_livox::LivoxHarmonicPlugin">
          <csv_file_name>$(find ros2_livox_simulation)/scan_mode/mid360.csv</csv_file_name>
          <topic>${topic}</topic>
          <samples>${samples}</samples>
          <downsample>1</downsample>
        </plugin>
      </sensor>
    </gazebo>
  </xacro:macro>

</robot>